openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +40,67 @@ def process_request(self, request):
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
             # what time is it now?
             utc_now = datetime.utcnow()
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    last_touch = datetime.fromisoformat(last_touch_str)
+                    time_since_last_activity = utc_now - last_touch
+
+                    has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds)
+                    if has_exceeded_timeout_limit:
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    log.warning("Parsing last touch time failed: %s", e)
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                # .. custom_attribute_name: session_inactivity.first_login
+                # .. custom_attribute_description: Tracks when users have no stored activity
+                #   timestamp (first access after login).
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
+                log.debug("No previous activity timestamp found (first login)")
+
+            cached_response = TieredCache.get_cached_response(LAST_SESSION_SAVE_TIME_KEYNAME)
+            last_save = cached_response.value if cached_response.is_found else None
+            current_time = datetime.utcnow().isoformat()
+
+            # .. custom_attribute_name: session_inactivity.activity_seen
+            # .. custom_attribute_description: Records the timestamp when user activity is detected.
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time)
+            if not last_save or (
+                datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < datetime.utcnow()
```

Comment:
1. Should we switch `datetime.utcnow()` to `current_time` for consistency?
2. To make my comment https://github.com/openedx/edx-platform/pull/36896#discussion_r2167654466 from the other day more explicit, I was saying you could replace comments with clear variable names, like the following:
```
has_save_delay_been_exceeded = datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < current_time
proceed_with_period_save = not last_save or has_save_delay_been_exceeded
if proceed_with_period_save:
   ...
```
Note: I switched from `datetime.utcnow()` to `current_time` again for consistency.